File: cracking-the-code-interview/chapter4/validate_bst.js
function BT(v) {
    this.root = this.node(v);
}

BT.prototype.node = v => ({
    val: v,
    left: null,
    right: null
});

BT.prototype.max = function (node = this.root, max = 0) {
//    if (!node) return max;
//
//    max = this.max(node.left, max);
//    max = Math.max(max, node.val);
//    max = this.max(node.right, max);
//
//    return max;
    let s = [node];

    while (s.length) {
        const n = s.pop();

        max = Math.max(max, n.val);

        if (n.left) s.push(n.left);
        if (n.right) s.push(n.right);
    }

    return max;
};

BT.prototype.min = function (node = this.root, min = this.root.val) {
    if (!node) return min;

    min = this.min(node.left, min);
    min = Math.min(min, node.val);
    min = this.min(node.right, min);

    return min;
};

BT.prototype.insert = function (v) {
    let node = this.root;

    while (node) {
        if (v < node.val) {
            if (!node.left) return node.left = this.node(v);
            node = node.left;
        } else {
            if (!node.right) return node.right = this.node(v);
            node = node.right;
        }
    }
};

const tree = new BT(50);

tree.insert(100);
tree.insert(75);
tree.insert(25);
tree.insert(90);
tree.insert(95);
tree.insert(60);
tree.insert(10);

/*
const validateBST = (node, v = []) => {
    if (!node) return v;

    validateBST(node.left, v);
    v.push(node.val);
    validateBST(node.right, v);

    return v;
};
*/

/*
const checkBST = () => {
    const nodes = validateBST(tree.root);

    for (let i = 0; i < nodes.length - 1; i++) {
        if (nodes[i] > nodes[i + 1]) return false;
    }

    return true;
};
*/

const checkBST = (node, last = -1) => {
    if (!node) return;

    checkBST(node.left, node.val);
    console.log(node.val, last);
    checkBST(node.right, node.val);
};

//const checkBST = tree =>
//    tree.max(tree.left) < tree.root.val < tree.min(tree.right);

console.log(checkBST(tree.root));

File: python/60-questions/tree-bt-bst/validate_bst.py
import ipdb


class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


# Recursive w/ valid range.
def validate_bst(root, low=float("-inf"), high=float("inf")):
    if not root:
        return True

    if not(low <= root.value <= high):
        return False

    return validate_bst(root.left, low, root.value) and validate_bst(root.right, root.value, high)


# Iterative w/ valid range.
def validate_bst(root):
    stack = [(root, float("-inf"), float("inf"))]

    while stack:
        node, low, high = stack.pop()

        if not(low <= node.value <= high):
            return False

        if node.left:
            stack.append((node.left, low, node.value))

        if node.right:
            stack.append((node.right, node.value, high))
    return True


# Iterative
def is_sorted(nums):
    for i in range(1, len(nums)):
        if nums[i] < nums[i - 1]:
            return False
    return True


# Iterative, inorder.
def validate_bst(root):
    stack = []
    current = root
    visited = []

    while current or stack:
        # Go left, young man.
        while current:
            stack.append(current)
            current = current.left

        current = stack.pop()
        visited.append(current.value)
        current = current.right
    return is_sorted(visited)


# Recursive, inorder.
def validate_bst(root, visited=[]):
    if not root:
        return visited

    validate_bst(root.left, visited)
    visited.append(root.value)
    validate_bst(root.right, visited)

    return is_sorted(visited)


# Iterative, compare to previous.
def validate_bst(root):
    stack = []
    current = root
    previous = float("-inf")

    while current or stack:
        while current:
            stack.append(current)
            current = current.left

        current = stack.pop()
        if previous > current.value:
            return False

        previous = current.value
        current = current.right
    return True


root = TreeNode(2)
root.left = TreeNode(1)
root.right = TreeNode(3)

root = TreeNode(5)
root.left = TreeNode(1)
root.right = TreeNode(7)
root.right.left = TreeNode(6)
root.right.right = TreeNode(8)

print(validate_bst(root))
